contracts/Registry.sol

```diff
@@ -1,70 +1,604 @@
+    }
+
+    function addTagsToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function addTagToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 tagName) internal {
+
+        // no-op if tag already exists
+        if (orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName == bytes32(0x0)) {
+            // add the service to the org level tag index
+            Tag memory tagObj;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName] = tagObj;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName = tagName;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].itemTagIndex = orgsByName[orgName].servicesByName[serviceName].tags.length;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].globalTagIndex = servicesByTag[tagName].orgNames.length;
+            orgsByName[orgName].servicesByName[serviceName].tags.push(tagName);
+
+            // add the service to the global tag index creating a list object for this tag if it does not already exist
+            if (!servicesByTag[tagName].valid) {
+                ServiceOrTypeRepositoryList memory listObj;
+                listObj.valid = true;
+                servicesByTag[tagName] = listObj;
+                serviceTags.push(tagName);
+            }
+
+            servicesByTag[tagName].orgNames.push(orgName);
+            servicesByTag[tagName].itemNames.push(serviceName);
+        }
+    }
+
+    function removeTagsFromServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        for (uint i = 0; i < tags.length; i++) {
+            removeTagFromServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function removeTagFromServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 tagName) internal {
+
+        // no-op if tag does not exist
+        if (orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName != bytes32(0x0)) {
+
+            // swap service registration lut entries
+            uint tagIndexToReplace = orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].itemTagIndex;
+            bytes32 tagNameToMove = orgsByName[orgName].servicesByName[serviceName].tags[orgsByName[orgName].servicesByName[serviceName].tags.length-1];
+
+            // no-op if we are deleting the last item
```

Comment:
Why don't we have this check in *all* of the places where we do this swap-and-shorten? (e.g. deleteOrganization, deleteServiceRegistrationInternal)

Reply:
will check whether I missed places.